docs/architecture/adr-008-event-subscription.md

```diff
@@ -0,0 +1,178 @@
+the `Subscription` owner's processing, it is proposed here to make use of an
+appropriate type of buffered channel to realize this. The size of this buffer
+must be configurable through the RPC Client.
+
+### RPC Pub/Sub
+
+The RPC Client must asynchronously spawn a component `PubSub` whose
+responsibilities are:
+
+1. Manage subscriptions (subscribe/unsubscribe operations from the RPC Client).
+2. Receive and deserialize incoming `Event`s.
+3. Route `Event`s to `Subscription`s whose `Query` matches the `Event`.
+4. Manage transport-level concerns (i.e. WebSockets connection: keep-alive
+   messages, graceful connection closure, etc.).
+
+Items (3) and (4) could be encapsulated into their own sub-components to allow
+for better separation of concerns and therefore easier testing. For example, (3)
+could be encapsulated into a `PubSubRouter` component and (4) into a component
+that only handles transport-level concerns.
+
+This is an *internal* component whose interface only matters to the RPC Client.
+
+### RPC Client Interface Extension
+
+The following interface is proposed as an extension to the existing [`Client`
+structure][client]:
+
+```rust
+impl Client {
+    /// Initiates a subscription for events produced by the given query. More
+    /// details as to the query syntax is available from the [Tendermint
+    /// `/subscribe` RPC endpoint documentation][tm-rpc-subscribe].
+    /// 
+    /// [tm-rpc-subscribe]: https://docs.tendermint.com/master/rpc/#/Websocket/subscribe
+    pub async fn subscribe(&mut self, query: &Query) -> Result<Subscription, Error> {
+        // ...
+    }
+
+    /// Terminates the given subscription and consumes it.
+    pub async fn unsubscribe(&mut self, subs: Subscription) -> Result<(), Error> {
+       // ...
+    }
+}
+```
+
+### Query
+
+Due to the potential complexity of queries, the proposed `Query` entity should
+functionally take its lead from the [Go-based `Query` interface][tm-go-query].
+This type of interface allows for simple short-term implementation (e.g. just
+querying by way of event type), with eventual full implementation of the [query
+PEG][query-peg].
+
+`Query` should implement `FromStr`, which should parse queries and emit the
+appropriate parsing errors:
+
+```rust
+let query = Query::from_str("tm.event='NewBlock'")?;
+```
+
+Matching of queries to events should also be simple:
+
+```rust
+if query.matches(event) {
+   // do something
+}
+```
+
+## Status
+
+Proposed
+
+## Consequences
+
+### Positive
+
+* Provides relatively intuitive developer ergonomics (`Subscription` iteration
+  to produce `Event`s).
+* Can be easily tested if the WebSockets transport layer is clearly encapsulated
+  from the perspective of the RPC Pub/Sub component.
+
+### Negative
+
+* Requires an additional concurrent, potentially long-running `async` task to be
+  concerned about.
+* Requires careful selection of `Subscription` buffer size to avoid lost events
```

Comment:
This could be a non-issue if we communicate clearly the guarantees of the client and/or actually have ways of tuning these.

Reply:
No longer relevant (yet), since I've taken out the bounded buffers from the architecture.